### PR TITLE
kubebuilder: init at 3.0.1

### PR DESCRIPTION
**Pull request by cmars**
###### Motivation for this change
kubebuilder is a common tool used for developing Kubernetes controllers.

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -1891,6 +1891,16 @@
     githubId = 3392199;
     name = "Calum MacRae";
   };
+  cmars = {
```
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:1894]: Please put this into a separate commit with the message ``maintainers: add NAME`` and sort it alphabetically.

File: pkgs/applications/networking/cluster/kubebuilder/default.nix [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+, go
+, git
+, cacert
+, make ? false }:
+
+buildGoModule rec {
+  pname = "kubebuilder";
+  version = "3.1.0";
+
+  src = fetchgit {
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/cluster/kubebuilder/default.nix:15]: Please use fetchFromGitHub.

File: pkgs/applications/networking/cluster/kubebuilder/default.nix [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      --replace "main.kubeBuilderVersion=" "main.kubeBuilderVersion=v${version}"
+  '';
+
+  buildPhase = ''
+    runHook preBuild
+    make build
+    mkdir -p $out/bin
+    cp bin/kubebuilder $out/bin
+    runHook postBuild
+  '';
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/cluster/kubebuilder/default.nix:36]: Please do not use the upstream makefile but nixpkgs buildGoModule logic and buildFlagsArray instead.

File: pkgs/applications/networking/cluster/kubebuilder/default.nix [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    runHook preCheck
+    make test-unit
+    runHook postCheck
+  '';
+
+  installPhase = ''
+    runHook preInstall
+    wrapProgram $out/bin/kubebuilder --set PATH ${lib.makeBinPath [ go ]}
+    runHook postInstall
+  '';
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/cluster/kubebuilder/default.nix:48]:
```suggestion
  postInstall = ''
    wrapProgram $out/bin/kubebuilder \
      --prefix PATH : ${lib.makeBinPath [ go ]}
  '';
```

File: pkgs/applications/networking/cluster/kubebuilder/default.nix [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+  buildInputs = [ go ];
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/cluster/kubebuilder/default.nix:50]:
```suggestion
```

File: pkgs/applications/networking/cluster/kubebuilder/default.nix [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  allowGoReference = true;
+  nativeBuildInputs = [ go git cacert make makeWrapper ];
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/cluster/kubebuilder/default.nix:52]:
```suggestion
  nativeBuildInputs = [ makeWrapper ];
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: maintainers/maintainer-list.nix (hunk 1 above)
Comment (cmars) [reply maintainers/maintainer-list.nix:1910]: Done.

File: pkgs/applications/networking/cluster/kubebuilder/default.nix (hunk 3 above)
Comment (cmars) [reply pkgs/applications/networking/cluster/kubebuilder/default.nix]: Done... but I'm still trying to figure out how to initialize these variables from `buildFlagsArray` (kubebuilder Makefile excerpt):

```
    -X main.gitCommit=$(shell git rev-parse HEAD) \
    -X main.buildDate=$(shell date -u +'%Y-%m-%dT%H:%M:%SZ') \
```

`main.gitCommit` should be doable (and reproducible), I just need to figure out how to evaluate a shell command into a build-time variable.

Perhaps `main.buildDate` just needs to be left out, or set to a fixed time for reproducibility? I can see how it's useful to the upstream developer to know when the binary was built, how does Nix generally approach these kinds of values in builds? Considering taking the timestamp of the git commit used above.

Comment (cmars) [reply pkgs/applications/networking/cluster/kubebuilder/default.nix]: Ok, figured it out, `pkgs/applications/virtualization/ignite/default.nix` was helpful.

File: pkgs/applications/networking/cluster/kubebuilder/default.nix [hunk 7]
```diff
@@ -0,0 +1,50 @@
+{ lib
+, buildGoModule
+, fetchFromGitHub
+, installShellFiles
+, makeWrapper
+, git
+, go ? false }:
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/cluster/kubebuilder/default.nix:7]:
```suggestion
, go
}:
```

File: pkgs/applications/networking/cluster/kubebuilder/default.nix [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  vendorSha256 = "0zxyd950ksjswja64rfri5v2yaalfg6qmq8215ildgrcavl9974n";
+
+  subPackages = ["cmd" "pkg/..."];
+
+  preBuild = ''
+    export buildFlagsArray+=("-ldflags=-X main.kubeBuilderVersion=v${version} \
+        -X main.goos=$GOOS \
+        -X main.goarch=$GOARCH \
+        -X main.gitCommit=$(git rev-parse HEAD) \
+        -X main.buildDate=$(git log -1 --date=iso-strict --format=%cd)")
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/cluster/kubebuilder/default.nix:30]: Can we set those to some variant of version to not need to fetch the .git directory?

Comment (cmars) [reply pkgs/applications/networking/cluster/kubebuilder/default.nix:30]: Hmm, I could see setting `gitCommit` to the tag (`v${version}`) since that basically references the commit. If I don't set buildDate though, it will default to the UNIX epoch:

```
[nix-shell:~/Projects/nixpkgs]$ kubebuilder version
Version: main.version{KubeBuilderVersion:"v3.1.0", KubernetesVendor:"unknown", GitCommit:"$Format:%H$", BuildDate:"1970-01-01T00:00:00Z", GoOs:"linux", GoArch:"amd64"}
```

I didn't want to set the buildDate to "now" in the build, because that seems like it'd break reproducibility.

I could see the commit hash and timestamp of that commit hash (as a proxy for actual build time) as being more useful to a kubebuilder developer receiving a bug report from someone who installed this package. But, I'm also new here, this is my first Nix package. Is it bad to rely on the .git directory in Nix builds? Is it possible to do a shallow fetch (--depth=1, etc) to minimize the overhead? What would you recommend?

**Comment by cmars** [comment]
@SuperSandro2000 Thanks for the review on this. As you can probably tell, this is my first Nix package. It's been a nice learning experience!

File: pkgs/applications/networking/cluster/kubebuilder/default.nix [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+}:
+
+buildGoModule rec {
+  pname = "kubebuilder";
+  version = "3.1.0";
+
+  src = fetchFromGitHub {
+    owner = "kubernetes-sigs";
+    repo = "kubebuilder";
+    url = "https://github.com/kubernetes-sigs/kubebuilder.git";
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/cluster/kubebuilder/default.nix:17]:
```suggestion
```

This is not required and is build together from owner and repo.

**Review by SuperSandro2000 — COMMENTED** [review]
other than that LGTM

File: pkgs/applications/networking/cluster/kubebuilder/default.nix (hunk 9 above)
Comment (cmars) [reply pkgs/applications/networking/cluster/kubebuilder/default.nix]: Removed
